docs/rfc/rfc-027-p2p-message-bandwidth-report.md

```diff
@@ -0,0 +1,137 @@
+
+#### Compress Transmitted Data
+
+Tendermint's data is sent uncompressed on the wire. The messages are not compressed before sending and the transport performs no compression either. Some of the information communicated by Tendermint is a poor candidate for compression: Data such as digital signatures and hashes have high entropy and therefore do not compress well. However, transactions may contain lots of information that has less entropy. Compression within Tendermint may be added at several levels. Compression may be performed at the [Tendermint 'packet' level][must-wrap-packet] or at the [Tendermint message send level][message-send].
+
+#### Transmit Less Data During Block Gossip
+
+Block, vote, and mempool gossiping transmit much of same data. The mempool reactor gossips candidate transactions to each peer. The consensus reactor, when gossiping the votes, sends vote metadata and the digital signature of that signs over that metadata. Finally, when a block is proposed, the proposing node amalgamates the received votes, a set of transaction, and adds a header to produce the block. This block is then serialized and gossiped as a list of bytes. However, the data that the block contains, namely the votes and the transactions were most likely _already transmitted to the nodes on the network_ via mempool transaction gossip and consensus vote gossip. 
+
+Therefore, block gossip can be updated to transmit a representation of the data contained in the block that assumes the peers will already have most of this data. Namely, the block gossip can be updated to only send 1) a list of transaction hashes and 2) a bit array of votes selected for the block along with the header and other required block metadata.
+
+This new proposed method for gossiping block data would require a slight update to the mempool transaction gossip and consensus vote gossip. Since all of the contents of each block will not be gossiped together, it's possible that some nodes are missing a proposed transaction or the vote of a validator indicated in the new block gossip format. The mempool and consensus reactors would need to be updated to provide a `NeedTxs` and `NeedVotes` message. Each of these messages would allow a node to request a set of data from their peers. When a node receives one of these, it will then transmit the Tx/Votes indicate in the associated message regardless of whether it believes it has transmitted them to the peer before.
```

Comment:
GossipSub works a little bit in this way with `Want` and `Have` messages. I think you'd want to have more than just `Want`/`Need` messages because you don't know who actually has the data you're requesting and so you're just blindly requesting it. What would be better is that upon receiving a block a node communicates all the txs it has. This can be quite compact in the same way `VoteSetBits` is because you reference by index in the block as opposed to hash

Reply:
This is a good approach to start thinking about, but I am afraid that the required changes are more complex. For instance, can a node forward a block when it does not have the full content (votes and txs) it references? Putting it in another way, how we guarantee that the references votes and txs are always available in the network?

Reply:
> Putting it in another way, how we guarantee that the references votes and txs are always available in the network?

[I mentioned this in another comment] Why do we need to guarantee that? How would that be different from forwarding a block part of a block for which we haven't received all block parts yet (in the current logic)?

Reply:
> I think you'd want to have more than just Want/Need messages because you don't know who actually has the data you're requesting and so you're just blindly requesting it.

This is true, however you the proposed mechanism still would result in the node receiving all of the data from its peers since our peers would have to receive them as well in order to actually commit the block at all.

> Putting it in another way, how we guarantee that the references votes and txs are always available in the network?

I'm a bit unclear on how this would be any different from what we have today. If a validator has already committed the block, then all of the Tx and votes within the block are available within the block. If the block has been pruned by all validators, then the Txs and votes may also be missing, but the block parts would also be gone so we are, as I see it, no worse than we previously were.